hpen.c

```diff
@@ -0,0 +1,491 @@
+		len--;
+		buf = (uint8_t *)buf + 1;
+	}
+
+	HPEN_FOREACH_USAGE_ABS(sc->abs_caps, usage) {
+		if (sc->locs_abs[usage].size > 0 && hpen_hid_map_abs[usage].code != HPEN_NO_CODE)
+			evdev_push_abs(sc->sc_evdev, hpen_hid_map_abs[usage].code,
+					hid_get_data(buf, len, &sc->locs_abs[usage]));
+	}
+
+	HPEN_FOREACH_USAGE_KEY(sc->key_caps, usage) {
+		if (sc->locs_key[usage].size > 0 && hpen_hid_map_key[usage].code != HPEN_NO_CODE)
+			evdev_push_key(sc->sc_evdev, hpen_hid_map_key[usage].code,
+					hid_get_data(buf, len, &sc->locs_key[usage]));
+	}
+
+	evdev_sync(sc->sc_evdev);
+}
+
+/* NOTE: in probe, called with sc==NULL just to check the pen collection exists */
+static int
+hpen_hid_parse(struct hpen_softc *sc, const void *d_ptr, uint16_t d_len,
+    uint32_t tlc_usage, uint8_t tlc_index)
+{
+	struct hid_item hi;
+	struct hid_data *hd;
+	size_t i;
+	bool pen_coll = false;
+	bool pen_found = false;
+
+	if (sc) {
+		bzero(sc->abs_caps, bitstr_size(HPEN_N_USAGES_ABS));
+		bzero(sc->key_caps, bitstr_size(HPEN_N_USAGES_KEY));
+	}
+
+	/* Parse inputs */
+	hd = hid_start_parse(d_ptr, d_len, 1 << hid_input);
+	HID_TLC_FOREACH_ITEM(hd, &hi, tlc_index) {
+		switch (hi.kind) {
+		case hid_collection:
+			if (hi.collevel == 1 && hi.usage ==
+			    HID_USAGE2(HUP_DIGITIZERS, HUD_PEN)) {
+				pen_coll = true;
+				pen_found = true;
+			}
+			break;
+		case hid_endcollection:
+			if (hi.collevel == 0 && pen_coll)
+				pen_coll = false;
+			break;
+		case hid_input:
+			if (!sc || !pen_coll)
+				break;
+			sc->report_id = hi.report_ID;
+			for (size_t i = 0; i < HPEN_N_USAGES_ABS; i++) {
+				if (hi.usage == hpen_hid_map_abs[i].usage) {
+					bit_set(sc->abs_caps, i);
+					sc->locs_abs[i] = hi.loc;
+					sc->ai[i] = (struct hpen_absinfo) {
+					    .max = hi.logical_maximum,
+					    .min = hi.logical_minimum,
+					    .res = hid_item_resolution(&hi),
+					};
+				}
+			}
+			for (size_t i = 0; i < HPEN_N_USAGES_KEY; i++) {
+				if (hi.usage == hpen_hid_map_key[i].usage) {
+					bit_set(sc->key_caps, i);
+					sc->locs_key[i] = hi.loc;
+				}
+			}
+			break;
+		default:
+			break;
+		}
+	}
+	hid_end_parse(hd);
+
+	if (!pen_found)
+		return (ENXIO);
+
+	if (!sc)
+		return (0);
+
+	for (i = 0; i < HPEN_N_USAGES_ABS; i++) {
+		if (hpen_hid_map_abs[i].required && !USAGE_SUPPORTED(sc->abs_caps, i)) {
+			device_printf(sc->sc_dev, "required report %s not found\n", hpen_hid_map_abs[i].name);
+			return (ENXIO);
+		}
+	}
+
+	for (i = 0; i < HPEN_N_USAGES_KEY; i++) {
+		if (hpen_hid_map_key[i].required && !USAGE_SUPPORTED(sc->key_caps, i)) {
+			device_printf(sc->sc_dev, "required report %s not found\n", hpen_hid_map_abs[i].name);
+			return (ENXIO);
+		}
+	}
+
+	sc->isize = hid_report_size_1(d_ptr, d_len, hid_input, sc->report_id);
+
+	return (0);
+}
+
+static int
+hpen_probe(device_t dev)
+{
+	struct hpen_softc *sc = device_get_softc(dev);
+	void *d_ptr;
+	uint16_t d_len;
+	int error;
+
+	if (hidbus_get_usage(dev) != HID_USAGE2(HUP_DIGITIZERS, HUD_PEN))
```

Comment:
It is better to use hidbus_lookup_id() or hidbus_lookup_driver_info() here. Static array of 'struct hid_device_id' type is intended to be exported for driver hotplug via devmatch utility. See MODULE_PNP_INFO(9)